### PR TITLE
Implement Hash#compact and Hash#compact!

### PR DESCRIPTION
**Pull request by KarolisL**

### CODE REVIEW

File: src/main/ruby/core/hash.rb [hunk 1]
```diff
@@ -183,6 +183,24 @@ def assoc(key)
     nil
   end
 
+  def compact
+    return Hash.new if empty?
```
Comment (eregon) [inline src/main/ruby/core/hash.rb]: `return {} if empty?` seems simpler here.
But this shortcut should not be needed, `reject` can handle empty Hash instances.

File: src/main/ruby/core/hash.rb [hunk 2]
```diff
@@ -183,6 +183,22 @@ def assoc(key)
     nil
   end
 
+  def compact
+    reject { |_, v| v.nil? }
+  end
+
+  def compact!
+    Truffle.check_frozen
```
Comment (eregon) [inline src/main/ruby/core/hash.rb]: Could this be just `reject!(&:nil?)`?

**Comment by chrisseaton** [comment]
Running CI now.

**Comment by eregon** [comment]
@KarolisL Could you rebase? Unfortunately spec/truffle.mspec is prone to conflicts in the current setup.

**Comment by eregon** [comment]
@KarolisL Thank you for the new feature!
Sorry it took so long to merge.

**Pull request merged**